From: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Date: Wed, 3 Feb 2016 09:49:18 +0530
Subject: [PATCH 14/28] be2net: Fix be_vlan_rem_vid() to check vlan id being
 removed
Patch-mainline: v4.6-rc1
Git-commit: 41dcdfbd0944a3ef08224a8c9ba3b2fdeae0dd86
References: bsc#963844 FATE#320192

The driver decrements its vlan count without checking if it is really
present in its list. This results in an invalid vlan count and impacts
subsequent vlan add/rem ops. The function be_vlan_rem_vid() should be
updated to fix this.

Signed-off-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/emulex/benet/be_main.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/net/ethernet/emulex/benet/be_main.c
+++ b/drivers/net/ethernet/emulex/benet/be_main.c
@@ -1463,6 +1463,9 @@ static int be_vlan_rem_vid(struct net_de
 	if (lancer_chip(adapter) && vid == 0)
 		return 0;
 
+	if (!test_bit(vid, adapter->vids))
+		return 0;
+
 	clear_bit(vid, adapter->vids);
 	adapter->vlans_added--;
 
